### PR TITLE
river: support "enum" blocks

### LINKED ISSUE #2602: River: support squashing structs into slices
This proposal extends #2373 to support squashing structs into slices. 

Shoutout to @tpaschalis for the help creating this proposal :) 

## Background 

Component authors may wish to support multiple variant blocks, but retain the order they are decoded in. This need was first identified in #2589, where users can provide one or more stage blocks to transform log lines, where stages are applied in the order they were defined in the config file. Unlike prometheus.relabel, there are multiple types of stages that users can use. 

Due to how the River API currently works, users must write stages like so:

```river 
loki.transform "example" {
  forward_to = [...]

  stage { 
    json { 
      ...
    }
  }
  stage {
    labels {
      ...
    } 
  }
}
```

This is verbosity being pushed to users because of the current limitations of the River API. It would be much more preferable for users to write the following:

```river 
loki.transform "example" {
  forward_to = [...]

  stage.json {
    ...
  }
  stage.labels {
    ...
  }
}
```

This proposal details changes to make to the River API to support decoding this while retaining the order of the blocks for proper processing of the stages. 

## Proposal 

I propose extending #2373 to permit squashing struct fields into slices. Component authors (like loki.process) could then define their arguments as:

```go 
type Arguments struct {
  ForwardTo []loki.LogsReceiver  `river:"forward_to,attr"`
  Stages    []StageConfig        `river:",squash"`
}

type StageConfig struct {
 JSON   *JSONStage   `river:"stage.json,block,optional"` 
 Labels *LabelsStage `river:"stage.labels,block,optional"`  
}
```

## Implementation 

When the River VM encounters a block which is defined by a squashed slice, the block is evaluated and decoded into a new instance of the struct, which is then appended to the slice.  This means that each squashed slice element will only ever have one field decoded. 

### No attributes 

To prevent users from being able to define attributes more than once, squashing slices is only supported when the struct type being sliced contains no attributes. For example, this is invalid:

```go
type Person struct {
  Attrs []Attributes `river:",squash"` 
}

type Attributes struct {
  // INVALID: Attribute is squashed as a slice into Person,
  // which would otherwise allow this strange config:
  //
  //   name = "John Doe"
  //   name = "Jane Doe" 
  Name string `river:"name,attr,optional"` 
}
``` 

The River VM must panic when given a squashed slice which breaks these rules. 

### Encoding 

For working with the River Value encoder, blocks within a slice of squashed structs must start with the same name fragment. 

This means that having blocks named `hello.world` and `hello.other.world` are valid, but `hello.world` and `fizz.buzz` are not. 

Given a slice of squashed structs of the following struct type:

```go
type StageConfig struct {
 JSON   *JSONStage   `river:"stage.json,block,optional"` 
 Labels *LabelsStage `river:"stage.labels,block,optional"`  
}
```

And a River config of:

```river 
stage.json {}
stage.labels {}
```

The raw value representation of the slice would look like the following:

```json 
{
  "stage": [
    {
      "json": {},
    },
    {
      "labels": {},
    },
  ],
}
``` 

River must panic when given an invalid squashed struct slice type.

## Other considerations 

The usage of this API will lead to a few scenarios: 

1. Users will almost always define blocks as pointers when squashing into a slice to make it easy to determine which block was decoded. 
2. It doesn't make sense for a user to mark a block as required if that block is being squashed into a slice. 

This proposal leaves these facts as-is, though in the future we may wish to do something about it (i.e., enforce usage of pointers, or enforce usage of optional).

### PR DESCRIPTION
**Pull request by rfratto**
#2602 originally proposed the idea of squashing structs into slices to support a slice where each element had exactly one field set to a non-zero value. 

I tried implementing it as originally suggested in #2602, but found it unintuitive and hard to follow in the code. 

This PR implements an alternative approach to #2602, supporting a new field type called "enum." Enum fields behave as described above: they're slices of a set of blocks, where each element in the slice will have exactly one inner block set. 

Enum fields must be given a name which will be prepended to the set of all blocks in the enum slice. 

For example, given the following two types:

```go 
type Arguments struct {
  Blocks []EnumBlock `river:"example,enum,optional"`  
}

type EnumBlock struct {
  BlockA struct{...} `river:"a,block,optional"`
  BlockB struct{...} `river:"b,block,optional"`
  BlockC struct{...} `river:"c,block,optional"`
}
```

Then the valid block names are `example.a`, `example.b`, and `example.c`. Any number of these blocks may be decoded from River, with the original order retained in the Blocks slice. 

This is a somewhat large PR, and I've tried to split up the work reasonably across multiple commits. I recommend reviewing commit-by-commit to make it easier. 

Closes #2602.

### CODE REVIEW

**Review by tpaschalis — APPROVED** [review]
LGTM! This was surprisingly easy to follow, thanks for making sure of that!

It does somewhat complicate the code in pkg/river, but I hope it's okay for the value it unlocks.

File: pkg/river/river.go [hunk 1]
```diff
@@ -86,6 +86,14 @@ import (
 //	// attributes and blocks of the outer struct.
 //	Field struct{...} `river:",squash"`
 //
+//	// Field appears as a set of blocks starting with "example.". Only the
+//	// first set element in the struct will be encoded. Each field in struct
+//	// must be a block. The name of the block is prepended to the enum name.
+//	Field []struct{...} `river:"example,enum"`
+//
+//	// Field is equivalent to `river:"example,enum"`.
+//	Field []struct{...} `river:"example,enum,optional"`
```
[the remark this answers is not in the document]
Comment (rfratto) [reply pkg/river/river.go:95]: I need to clarify the docs here; optional is technically enforced, and there's no way to make it required. 

I think we might change the semantics of being required in the future (when we have time), so you should still flag it as optional if the intent is for it to be optional.

File: pkg/river/encoding/block.go [hunk 2]
```diff
@@ -86,23 +86,80 @@ func getFieldsForBlock(input interface{}) ([]interface{}, error) {
...
+		if enumElement.IsNil() {
+			return nil, nil
+		}
+		enumElement = enumElement.Elem()
+	}
+
+	fields := rivertags.Get(enumElement.Type())
+
+	// Find the first non-zero field and encode it as a block.
+	for _, field := range fields {
```
[the remark this answers is not in the document]
Comment (tpaschalis) [reply pkg/river/encoding/block.go:148]: disregard that, I misread the code here.
